File: Backend/src/main/java/com/example/backend/entity/Quiz.java
package com.example.backend.entity;

import com.example.backend.dto.QuizDTO;
import jakarta.persistence.*;

import java.util.List;

@Entity
@Table(name = "quiz")
public class Quiz {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    // Other attributes
    private String title;
    private String subject;
    private int time;
    @ManyToMany(cascade = { CascadeType.PERSIST, CascadeType.MERGE })
    @JoinTable(
            name = "quiz_question",
            joinColumns = @JoinColumn(name = "quiz_id"),
            inverseJoinColumns = @JoinColumn(name = "question_id")
    )
    private List<Question> questions;

    @OneToMany(mappedBy = "quiz", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Test> tests;

    @ManyToOne(cascade = CascadeType.MERGE, fetch= FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;


    // Constructors, getters, and setters
    // ...

    public Quiz() {
    }

    public Quiz(int id, String title, String subject, int time, List<Question> questions, List<Test> tests, User user) {
        this.id = id;
        this.title = title;
        this.subject = subject;
        this.time = time;
        this.questions = questions;
        this.tests = tests;
        this.user = user;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getSubject() {
        return subject;
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

    public int getTime() {
        return time;
    }

    public void setTime(int time) {
        this.time = time;
    }

    public List<Question> getQuestions() {
        return questions;
    }

    public void setQuestions(List<Question> questions) {
        this.questions = questions;
    }

    public List<Test> getTests() {
        return tests;
    }

    public void setTests(List<Test> tests) {
        this.tests = tests;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }
}

File: front_end/lib/app/core/models/quiz.dart
import 'question.dart';
import 'test.dart';

class Quiz {
  final int id;
  final String subject;
  final String title;
  final int time;
  final List<Question> questions;
  final List<Test> tests;

  Quiz({
    required this.id,
    required this.subject,
    required this.title,
    required this.time,
    required this.questions,
    required this.tests,
  });

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'subject': subject,
      'title': title,
      'time': time,
      'questions': questions.map((x) => x.toJson()).toList(),
      'tests': tests.map((x) => x.toJson()).toList(),
    };
  }

  factory Quiz.fromJson(Map<String, dynamic> map) {
    return Quiz(
      id: map['id'] as int,
      subject: map['subject'] as String,
      title: map['title'] as String,
      time: map['time'] as int,
      questions: List<Question>.from(
        (map['questions'] as List).map<Question>(
          (x) => Question.fromJson(x as Map<String, dynamic>),
        ),
      ),
      tests: List<Test>.from(
        (map['tests'] as List).map<Test>(
          (x) => Test.fromJson(x as Map<String, dynamic>),
        ),
      ),
    );
  }

  Quiz copyWith({
    int? id,
    String? subject,
    String? title,
    int? time,
    List<Question>? questions,
    List<Test>? tests,
  }) {
    return Quiz(
      id: id ?? this.id,
      subject: subject ?? this.subject,
      title: title ?? this.title,
      time: time ?? this.time,
      questions: questions ?? this.questions,
      tests: tests ?? this.tests,
    );
  }
}
